### PR TITLE
feat: add support for set() with SetOptions when using FirestoreDataConverter

### LINKED ISSUE #1000: WriteBatch.set should have a signature override
`WriteBatch.set` changes its semantics when passing options, specially when providing `{merge: true}`

In that case an overload to the signature should be added, as in that case there's no need to provide a full document.

The [original definition](c89546f5ae83da3845076aeeffcda75f9b208f5c) is:

```ts
set<T>(documentRef: DocumentReference<T>, data: T, options?: SetOptions): WriteBatch;
``` 

The proposed overload would be something like:
```ts
set<T>(documentRef: DocumentReference<T>, data: Partial<T>, {merge: true}): WriteBatch;
```

### PR DESCRIPTION
**Pull request by thebrianchen**
Fixes #1000.

### CODE REVIEW

File: dev/src/write-batch.ts [hunk 1]
```diff
@@ -296,14 +306,23 @@ export class WriteBatch {
...
   ): WriteBatch {
     validateSetOptions('options', options, {optional: true});
     const mergeLeaves = options && options.merge === true;
     const mergePaths = options && options.mergeFields;
     validateDocumentReference('documentRef', documentRef);
-    let firestoreData = documentRef._converter.toFirestore(data);
+    let firestoreData: DocumentData;
+    if (mergeLeaves || mergePaths) {
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      firestoreData = (documentRef._converter as any).toFirestore(
```
Comment (schmidt-sebastian) [inline dev/src/write-batch.ts:319]: Please add a short comment on why we need this cast to `any`.

File: dev/test/util/helpers.ts [hunk 2]
```diff
@@ -342,6 +342,19 @@ export const postConverter = {
   },
 };
 
+export const postConverterMerge = {
+  toFirestore(post: Partial<Post>): DocumentData {
```
Comment (schmidt-sebastian) [inline dev/test/util/helpers.ts:346]: Could we add SetOptions here and additional asserts? Something like that:

```
if (merge) {
expect(post).to.be.an.instanceOf(Post);
} else {
expect(post).to.not.be.an.instanceOf(Post);
}
```

File: dev/src/document.ts [hunk 3]
```diff
@@ -382,12 +388,14 @@ export class DocumentSnapshot<T = DocumentData> {
...
     // We only want to use the converter and create a new QueryDocumentSnapshot
     // if a converter has been provided.
-    if (this.ref._converter !== defaultConverter) {
+    if (this.ref._converter !== defaultConverter()) {
       const untypedReference = new DocumentReference(
         this.ref.firestore,
         this.ref._path
       );
-      return this.ref._converter.fromFirestore(
+      return ((this.ref._converter as unknown) as FirestoreDataConverter<
```
Comment (schmidt-sebastian) [inline dev/src/document.ts:396]: This compiles for me as just "this.ref._converter.fromFirestore".

Comment (thebrianchen) [reply dev/src/document.ts:396]: done.

File: dev/test/util/helpers.ts (hunk 2 above)
Comment (thebrianchen) [reply dev/test/util/helpers.ts:346]: Done, but flipped the expect statements.

File: dev/src/write-batch.ts (hunk 1 above)
Comment (thebrianchen) [reply dev/src/write-batch.ts:321]: done.

**Comment by schmidt-sebastian** [comment]
Do you want to make this a "feat" before merging? This deserves more recognition :)

**Pull request merged**